### DISCUSSION TITLE
Comments for `AbortMultipartUpload` inconsistent with the behavior

### DISCUSSION THREAD

**Discussion by JinnyYi** [minio/minio#13495] [Q&A]
As stated in [the comments of AbortMultipartUpload](https://github.com/minio/minio/blob/master/cmd/fs-v1-multipart.go#L801-L804), it is an atomic idempotent operation. But it looks inconsistent with [the implementation](https://github.com/minio/minio/blob/master/cmd/fs-v1-multipart.go#L823-L831) and test results.

## Expected Behavior

Code comments should be consistent with the implementation.

- If `AbortMultipartUpload` is idempotent, then it should not return `NoSuchUpload`.
- If `AbortMultipartUpload` is non-idempotent, maybe the comments should be updated. 

## Current Behavior

The comment of `AbortMultipartUpload` is inconsistent with the implementation:

`AbortMultipartUpload`: [Implements S3 compatible Abort multipart API, slight difference is that this is an atomic idempotent operation](https://github.com/minio/minio/blob/master/cmd/fs-v1-multipart.go#L801-L804). But when we call it multiple times to abort a multi part upload with the upload ID, we get the `NoSuchUpload` error.

![minio](https://user-images.githubusercontent.com/82445294/138386258-bbfe1ec7-2146-4f81-ad8c-632e69c7a328.png)

## Steps to Reproduce (for bugs)

Testing compatibility with S3:

1. `CreateMultipart` to initiate a multipart upload and get the upload ID.
2. The first time `AbortMultipartUpload` with the upload ID is called,  it returns success.
3. The second time `AbortMultipartUpload` with the upload ID is called,  it returns `NoSuchUpload`.

## Context

We define the behavior of [Idempotent Storager Delete Operation](https://github.com/beyondstorage/specs/blob/master/rfcs/46-idempotent-delete.md), if `AbortMultipartUpload` is non-idempotent in MinIO, we will ignore `NoSuchUplaod` error in our [services/s3](https://github.com/beyondstorage/go-storage/tree/master/services/s3) for compatible services.

## Your Environment

- MinIO: minio version RELEASE.2021-10-13T00-23-17Z
- aws s3: aws-sdk-go-v2 v1.9.2

**Comment by harshavardhana** [comment]
Not sure what Is an issue here, moving this to the discussion.

**Comment by harshavardhana** [comment]
Can you explain more clearly? - are you asking us to behave like AWS S3?

**Reply by JinnyYi** [reply]
I want to confirm the behavior of `AbortMultipartUpload`, whether it is idempotent:

- If `AbortMultipartUpload` is idempotent, then maybe it should not return `NoSuchUpload` even the multipart upload has been aborted.
- If `AbortMultipartUpload` is non-idempotent, maybe [the comments](https://github.com/minio/minio/blob/master/cmd/fs-v1-multipart.go#L801-L804) should be updated.

**Reply by harshavardhana** [reply]
> * If `AbortMultipartUpload` is idempotent, then maybe it should not return `NoSuchUpload` even the multipart upload has been aborted.

On MinIO it is idempotent.

**Reply by JinnyYi** [reply]
On AWS S3, `AbortMultipartUpload` with the same upload id will always get `204 No Content`.
But on MinIO, when the client repeats the same `AbortMultipartUpload` request, as the resource has already been deleted, the server returns `404 Not Found`.

We expect the behavior to be consistent with AWS S3.

**Reply by harshavardhana** [reply]
> On AWS S3, `AbortMultipartUpload` with the same upload id will always get `204 No Content`.
> But on MinIO, when the client repeats the same `AbortMultipartUpload` request, as the resource has already been deleted, the server returns `404 Not Found`.

AWS S3 is inconsistent in its behavior, it returns 404 after 24hours - we are strictly consistent object storage and so our behavior is more inline with expectation.

**Reply by Xuanwo** [reply]
Oh, that's new to us. We will re-check out logic around the `AbortMultipartUpload` related logic.

BTW, do you think it's a good idea to add this behavior detail to minio's comments? Because we never thought it would be for this reason.

---

- Problem Issus: https://github.com/beyondstorage/go-storage/issues/860
- Our Fix: https://github.com/beyondstorage/go-storage/pull/937#pullrequestreview-786622228

**Reply by harshavardhana** [reply]
> BTW, do you think it's a good idea to add this behavior detail to minio's comments? Because we never thought it would be for this reason.

There is nothing to comment really - MinIO's behavior is strongly consistent it is mentioned in the docs 

- https://docs.min.io/docs/distributed-minio-quickstart-guide.html
- https://github.com/minio/minio/tree/master/docs/distributed#consistency-guarantees
